File: fastfood.py
#!/usr/bin/env python
from theano import function
import theano.tensor as T
from theano.tensor import shared_randomstreams
import numpy as np
import numpy.random
from scipy.special import gammaincinv
from numpy.linalg import norm

# tensor stand-in for np.random.RandomState
rngT = shared_randomstreams.RandomStreams()
rng = numpy.random.RandomState()

# {{{ Fastfood Params }}}
n, d = T.dscalars('n', 'd')
# transform dimensions to be a power of 2
d0, n0 = d, n
l = T.ceil(T.log2(d))  # TODO cast to int
d = 2**l
k = T.ceil(n/d)  # TODO cast to int
n = d*k
# generate parameter 'matrices'
B = rng.choice([-1, 1], size=(k, d))
G = rng.normal(size=(k, d), dtype=np.float64)
PI = np.array([rng.permutation(d) for _ in xrange(k)]).T
S = np.empty((k*d, 1), dtype=np.float64)
# generate scaling matrix, S
for i in xrange(k):
    for j in xrange(d):
        p1 = rng.uniform(size=d)
        p2 = d/2
        Tmp = gammaincinv(p2, p1)
        Tmp = T.sqrt(2*Tmp)
        s_ij = Tmp * norm(G, 'fro')**(-1)
        S[i+j] = s_ij
fastfood_params = function([n, d], [B, G, PI, S])


# {{{ Fastfood for kernel }}}
# params
X = T.dmatrix('X')  # COLUMNS are input patterns (d0 dims, m samples)
B = T.dmatrix('B')
G = T.dmatrix('G')
PI = T.dmatrix('PI')
S = T.dmatrix('S')
para = [B, G, PI, S]
sgm = T.dscalar('sgm')
# book keeping
d0, m = X.shape
l = T.ceil(T.log2(d0))  # TODO cast to int
d = 2**l
if d == d0:
    XX = X
else:
    XX = np.zeros((d, m))
    XX[0:d0, :] = X
#
k = len(para.B)
n = d*k
tht = np.zeros((n, m))
XX = B*XX
Tmp = FWHT(XX)
Tmp = Tmp[PI]  # TODO make this indexing work as expected
Tmp = G*d*Tmp
Tmp = FWHT(T)
tht[ii*d:(ii+1)*d, :] = Tmp
fastfood_forkernel = function([X, para, sgm], [phi, tht])

File: fastfood.R
fastfood.params <- function(n, d) {
  d0 <- d
  n0 <- n
  l <- as.integer(ceiling(log2(d)))
  d <- 2^l
  k <- as.integer(ceiling(n/d))
  n <- d*k
  print(paste('d0=',d0,', d=',d))
  print(paste('n0=',n0,', n=',n))
  
  B <- matrix(nrow=k,ncol=d)
  G <- matrix(nrow=k,ncol=d)
  PI <- matrix(nrow=k,ncol=d)
  S <- matrix(nrow=k,ncol=d)
  for(ii in 1:k) {
    B[ii,] <- sample(c(-1,1),size=d,replace=T)
    G[ii,] <- rnorm(d)
    PI[ii,] <- sample.int(d,d)
    
    p1 <- runif(d)
    p2 <- d / 2
    Tmp <- sqrt(2*gammaincinv(p1, p2))
    s.i <- Tmp*(1/norm(G,'F'))
    S[ii,] <- s.i
  }
  
  params <- data.frame(B=B,G=G,PI=PI,S=S)
  return(params)
}

fastfood.forkernel <- function(X,para,sgm) {
  c(d0, m) <- dim(X)
  l <- as.integer(ceiling(log2(d0)))
  d <- 2^l
  if (d == d0) {
    XX <- X
  } else {
    XX <- matrix(nrow=d,ncol=m)
    XX[,] <- 0
    XX[1:d0,] <- X
  }
  k <- nrow(para$B)
  n <- d*k
  tht <- matrix(nrow=n,ncol=m)
  tht[,] <- 0
  for (ii in 1:k) {
    B <- diag(para$B[ii,],nrow=d,ncol=d)
    G <- diag(para$G[ii,],nrow=d,ncol=d)
    PI <- diag(para$PI[ii,],nrow=d,ncol=d)
    xx <- B%*%XX
    Tmp <- fwht(xx)
    Tmp <- Tmp[PI,]
    Tmp <- (d*G)%*%Tmp
    Tmp <- fwht(Tmp)
    idx1 <- (ii-1)*d
    idx2 <- ii*d
    tht[idx1:idx2,] <- Tmp
  }
  tht <- t((S*sqrt(d))*t(tht))
  Tmp <- tht/sgm
  #line 218
}